pkg/policyfilter/map.go

```diff
@@ -95,36 +177,109 @@ func (m PfMap) newPolicyMap(polID PolicyID, cgIDs []CgroupID) (polMap, error) {
 	}
 
 	// update inner map with ids
-	ret := polMap{inner}
+	ret := polMap{
+		Inner:   inner,
+		Reverse: m.rev,
+	}
 	if err := ret.addCgroupIDs(cgIDs); err != nil {
-		ret.Close()
+		ret.Inner.Close()
 		return polMap{}, fmt.Errorf("failed to update policy (id=%d): %w", polID, err)
 	}
 
 	// update outer map
 	// NB(kkourt): use UpdateNoExist because we expect only a single policy with a given id
-	if err := m.Update(polID, uint32(ret.FD()), ebpf.UpdateNoExist); err != nil {
-		ret.Close()
+	if err := m.dir.Update(polID, uint32(ret.Inner.FD()), ebpf.UpdateNoExist); err != nil {
+		ret.Inner.Close()
 		return polMap{}, fmt.Errorf("failed to insert inner policy (id=%d) map: %w", polID, err)
 	}
 
+	// update reverse map
+	for _, cgID := range cgIDs {
+		if err := addReverseMapping(m.rev, polID, cgID); err != nil {
+			return polMap{}, fmt.Errorf("failed to update reverse map: %w", err)
+		}
+	}
+
 	return ret, nil
 }
 
-func (m PfMap) readAll() (map[PolicyID]map[CgroupID]struct{}, error) {
+func getMapSize(m *ebpf.Map) (uint32, error) {
+	var key uint32
+	var val uint8
+	var mapSize uint32
+
+	inIter := m.Iterate()
+	for inIter.Next(&key, &val) {
+		mapSize++
+	}
+
+	if err := inIter.Err(); err != nil {
+		return 0, fmt.Errorf("error iterating inner map: %w", err)
+	}
+
+	return mapSize, nil
+}
+
+func (m PfMap) deletePolicyIdInReverse(polID PolicyID) error {
+	var key CgroupID
+	var id uint32
+
+	cgIDs := []CgroupID{}
+	iter := m.rev.Iterate()
+	for iter.Next(&key, &id) {
+		inMap, err := ebpf.NewMapFromID(ebpf.MapID(id))
+		if err != nil {
+			return fmt.Errorf("error opening inner map: %w", err)
+		}
+		defer inMap.Close()
+
+		// We don't know if this exists if this does not exist this
+		// will return an error, but this is fine.
+		inMap.Delete(polID)
+
+		// now we need to check the size of the inner map
+		// if this is 0 we should also remove the outer entry
+		mapSize, err := getMapSize(inMap)
+		if err != nil {
+			return fmt.Errorf("error getting inner map size: %w", err)
+		}
+		if mapSize == 0 {
+			cgIDs = append(cgIDs, key)
+		}
+	}
+
+	if err := iter.Err(); err != nil {
+		return fmt.Errorf("deletePolicyIdInReverse: error iterating reverse map: %w", err)
```

Comment:
if you change, don't forget to change here.

Reply:
Same here.